File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int amountOfTime(TreeNode root, int start) {
        List<Integer> result = new ArrayList<>();
        result.add(0);
        traverse(root, start, result);
        return result.get(0);
    }
    
    public int[] traverse(TreeNode root, int start, List<Integer> result) {
        if (root == null) return new int[]{0, 0};
        int[] left = traverse(root.left, start, result);
        int[] right = traverse(root.right, start, result);
        if (left[1] == -1) {
            if (left[0]> result.get(0)) {
                result.set(0, left[0]);
            }
            if (right[0] != 0 && right[0] + left[0]> result.get(0)) {
                result.set(0, right[0] + left[0]);
            }
            return new int[]{left[0] + 1, -1};
        } else if (right[1] == -1) {
            if (right[0]> result.get(0)) {
                result.set(0, right[0]);
            }
            if (left[0] != 0 && right[0] + left[0]> result.get(0)) {
                result.set(0, right[0] + left[0]);
            }
            return new int[]{right[0] + 1, -1};
        } else {
            int max_len = Math.max(left[0], right[0]);
            if (root.val == start) {
                result.set(0, max_len);
                return new int[]{1, -1};
            }
            return new int[]{max_len + 1, 0};
        }
    }
}

File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        # if the infection happens on the left subtree
        # and the infection happens on the right subtree
        max_time = float("-inf")
        def infect(root):
            nonlocal max_time
            # return (infected, height)
            if root is None:
                return (False, 0, 0)
            current_val = root.val
            is_current_start = False
            if current_val == start:
                # infected node
                is_current_start = True
            left_infected, left_height, left_delta = infect(root.left)
            right_infected, right_height, right_delta = infect(root.right)
            if not left_infected and not right_infected and is_current_start:
                max_time = max(max_time, left_height, right_height)
                return (True, 0, 1)
            if left_infected:
                max_time = max(max_time, left_delta + right_height)
                return (True, 0, left_delta + 1)
            elif right_infected:
                max_time = max(max_time, right_delta + left_height)
                return (True, 0, right_delta + 1)
            return (False, 1 + max(left_height, right_height), 0)
        infect(root)
        return max_time
